File: tests/workflow/producer/rust/src/producer/implementation/beacons/beacons_shutdownserver.rs
use super::{
    beacons, identification, pack, producer::Control, protocol, Context, EmitterError,
    ProducerError,
};
use clibri::server;

pub async fn emit<E: server::Error, C: server::Control<E> + Send + Clone>(
    identification: &mut identification::Identification,
    beacon: &protocol::Beacons::ShutdownServer,
    sequence: u32,
    filter: &identification::Filter,
    context: &mut Context,
    control: &Control<E, C>,
) -> Result<(), EmitterError> {
    beacons::beacons_shutdownserver::emit::<E, C>(identification, beacon, filter, context, control)
        .await
        .map_err(EmitterError::Emitting)?;
    let mut response = protocol::InternalServiceGroup::BeaconConfirmation { error: None };
    let buffer = pack(&sequence, &identification.uuid(), &mut response)?;
    control
        .send(buffer, Some(identification.uuid()))
        .await
        .map_err(|e: ProducerError<E>| EmitterError::Emitting(e.to_string()))?;
    Ok(())
}

File: tests/workflow/producer/typescript/src/producer/implementation/beacons/beacons.shutdownserver.ts
import { Producer, Identification, Filter, Context, Protocol } from "./index";
import { emit } from "../../beacons/beacons.shutdownserver";
import { Scope } from "../scope";

export function handler(
    beacon: Protocol.Beacons.ShutdownServer,
    consumer: Identification,
    filter: Filter,
    context: Context,
    producer: Producer,
    sequence: number
): Promise<void> {
	const scope = new Scope(consumer, filter, context, producer);
	return new Promise((resolve, reject) => {
		emit(beacon, scope)
			.then(() => {
				const confirmation =
					new Protocol.InternalServiceGroup.BeaconConfirmation({
						error: undefined,
					});
				producer
					.send(
						consumer.uuid(),
						confirmation.pack(sequence, consumer.uuid())
					)
					.then(() => {
						scope.call();
						resolve();
					})
					.catch(reject);
			})
			.catch(reject);
	});
}